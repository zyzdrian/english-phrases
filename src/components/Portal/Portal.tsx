import React, { useEffect, useState } from 'react';
import ReactDOM from 'react-dom';
import { addRootElement, createRootElement } from '../../utils';

export const Portal: React.FC<IPortal> = ({ id, children }) => {
    const [el] = useState(document.createElement('div'));
    useEffect(() => {
        const existingParent = document.getElementById(id);
        const parentElem: HTMLElement = existingParent || createRootElement(id);
        parentElem.setAttribute('data-testid', `portal-${id}`);

        if (!existingParent) {
            addRootElement(parentElem);
        }

        parentElem.appendChild(el);
        return () => {
            el.remove();
            if (!parentElem.childNodes.length) {
                parentElem.remove();
            }
        };
    }, []);
    return ReactDOM.createPortal(children, el);
};

export interface IPortal {
    id: string;
}