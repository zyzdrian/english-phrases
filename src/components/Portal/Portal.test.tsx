import React from 'react';
import { cleanup, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Portal, IPortal } from './Portal';

describe('<Portal /> tests', () => {
    const INITIAL_PROPS: IPortal = {
        id: 'test-root'
    };
    const children = <div>TEST</div>;
    let props = { ...INITIAL_PROPS };
    beforeEach(() => {
        props = { ...INITIAL_PROPS };
    });
    afterEach(cleanup);

    it('should create portal with children', () => {
        const { container, getByTestId } = render(<Portal {...props}>{children}</Portal>);
        expect(container.firstChild).toMatchSnapshot();
        expect(getByTestId(`portal-${props.id}`)).toBeTruthy();
    });

    it('should remove children from portal', () => {
        const { container, unmount, queryByTestId } = render(<Portal {...props}>{children}</Portal>);
        unmount();

        expect(container.firstChild).toMatchSnapshot();
        expect(queryByTestId(`portal-${props.id}`)).toBeNull();
    });

    it('shouldn\'t remove parent element', () => {
        const { container, unmount, queryByTestId } = render(<Portal {...props}>{children}</Portal>);
        render(<Portal {...props}>{children}</Portal>);
        unmount();

        expect(container.firstChild).toMatchSnapshot();
        expect(queryByTestId(`portal-${props.id}`)).toBeTruthy();
    });
});
