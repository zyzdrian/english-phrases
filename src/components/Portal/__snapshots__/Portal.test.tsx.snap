// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portal /> tests should create portal with children 1`] = `null`;

exports[`<Portal /> tests should remove children from portal 1`] = `null`;

exports[`<Portal /> tests shouldn't remove parent element 1`] = `null`;
